src/algorithms/policy_gradient/vpg.jl

```diff
@@ -0,0 +1,140 @@
+    batch_size::Int = 1024
+    rng::R = Random.GLOBAL_RNG
+    loss::Float32 = 0.0f0
+    baseline_loss::Float32 = 0.0f0
+end
+
+"""
+About continuous action space, see
+* [Diagonal Gaussian Policies](https://spinningup.openai.com/en/latest/spinningup/rl_intro.html#stochastic-policies
+* [Clipped Action Policy Gradient](https://arxiv.org/pdf/1802.07564.pdf)
+"""
+
+function (π::VPGPolicy)(env::AbstractEnv)
+    to_dev(x) = send_to_device(device(π.approximator), x)
+
+    logits = env |> get_state |> to_dev |> π.approximator
+
+    if π.action_space isa DiscreteSpace
+        dist = logits |> softmax |> π.dist
+        action = π.action_space[rand(π.rng, dist)]
+    elseif π.action_space isa ContinuousSpace
+        dist = π.dist.(logits...)
+        action = rand.(π.rng, dist)[1]
+    else
+        error("not implemented")
+    end
+    action
+end
+
+function (π::VPGPolicy)(env::MultiThreadEnv)
+    error("not implemented")
+    # TODO: can PG support multi env? PG only get updated at the end of an episode.
+end
+
+function RLBase.update!(π::VPGPolicy, traj::ElasticCompactSARTSATrajectory)
+    (length(traj[:terminal]) > 0 && traj[:terminal][end]) || return
+
+    model = π.approximator
+    to_dev(x) = send_to_device(device(model), x)
+
+    states = traj[:state]
+    actions = traj[:action] |> Array # need to convert ElasticArray to Array, or code will fail on gpu. `log_prob[CartesianIndex.(A, 1:length(A))`
+    gains = traj[:reward] |> x -> discount_rewards(x, π.γ)
+
+    for idx in Iterators.partition(shuffle(1:length(traj[:terminal])), π.batch_size)
+        S = select_last_dim(states, idx) |> to_dev
+        A = actions[idx]
+        G = gains[idx] |> x -> Flux.unsqueeze(x, 1) |> to_dev
+        # gains is a 1 colomn array, but the ouput of flux model is 1 row, n_batch columns array. so unsqueeze it.
+
+        if π.baseline isa NeuralNetworkApproximator
+            gs = gradient(Flux.params(π.baseline)) do
+                δ = G - π.baseline(S)
+                loss = mean(δ .^ 2) * π.α_w # mse
+                ignore() do
+                    π.baseline_loss = loss
+                end
+                loss
+            end
+            update!(π.baseline, gs)
+        elseif π.baseline isa Nothing
+            # Normalization. See
+            # (http://rail.eecs.berkeley.edu/deeprlcourse-fa17/f17docs/hw2_final.pdf)
+            # (https://web.stanford.edu/class/cs234/assignment3/solution.pdf)
+            # normalise should not be used with baseline. or the loss of the policy will be too small.
+            δ = G |> x -> normalise(x; dims = 2)
+        end
+
+        gs = gradient(Flux.params(model)) do
+            if π.action_space isa DiscreteSpace
+                log_prob = S |> model |> logsoftmax
+                log_probₐ = log_prob[CartesianIndex.(A, 1:length(A))]
+            elseif π.action_space isa ContinuousSpace
+                dist = π.dist.(model(S)...) # TODO: Normal. does not work on GPU. InvalidIRError.
```

Comment:
Just took a glance at your implementation as this might be interesting for my project too.

I actually had the same issue with the Soft-Actor-Critic (the reason the implementation in this framework is not GPU capable). There seems to be an issue with drawing random numbers on the whole CUDA + Zygote stack. Unfortunatly, I was not able to really pinpoint which function was responsible for for the error.

However, I resolved this issue by implementing my own loglikelihood for normal distributions:

```
const log2π = log(2f0 * Float32(π))

function my_loglikelihood(μ, σ, x)
    z = (x .- μ) ./ σ
    -(z .^ 2 .+ log2π) ./ 2f0
end
```

Maybe this solves your issue too?

Reply:
I filed an issue for `Distributions`. https://github.com/JuliaStats/Distributions.jl/issues/1183
It seems the problem is related to CUDA, and can not be solved by using `DistributionsAD`.

I think your code is a better solution. Shall we put this code in somewhere, such as `ReinforcementLearningCore/src/utils/base.jl`?

I guess there is a typo in `log2π`?
```
function logpdf1(μ, σ, x)
    z = (x .- μ) ./ σ
    -(z .^ 2 .+ log(2π)) ./ 2f0
end
```

Reply:
Nice, maybe your [issue](https://github.com/JuliaStats/Distributions.jl/issues/1183) will raise some attention towards that topic.

Im not really comfortable with using this kind of 'hacks' in a frameworks because it basically creates redundant code, but yes it would help a lot. Maybe @findmyway can comment on finding a good place for such patches?

Regarding my implementation:
I defined an constant for `log2π` to avoid Float64 creeping into this, but something like `log(2f0π)` will work too of course!

Reply:
But for the sake of merging this PR, I think it might be best to fix this in at a later point in a new PR?

Reply:
I usually put this kind of patches into [ReinforcementLearningCore.jl/tree/master/src/extensions](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/tree/master/src/extensions) and try to merge them into upstream packages when possible.